control-plane/connect-inject/endpoints_controller.go

```diff
@@ -122,86 +123,106 @@ type EndpointsController struct {
 	context.Context
 }
 
-// Reconcile reads the state of an Endpoints object for a Kubernetes Service and reconciles Consul services which
+// Reconcile reads the state of an EndpointSlice object for a Kubernetes Service and reconciles Consul services which
 // correspond to the Kubernetes Service. These events are driven by changes to the Pods backing the Kube service.
 func (r *EndpointsController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	var errs error
-	var serviceEndpoints corev1.Endpoints
+	var serviceEndpointSlice v1.EndpointSlice
 
 	// Ignore the request if the namespace of the endpoint is not allowed.
 	if shouldIgnore(req.Namespace, r.DenyK8sNamespacesSet, r.AllowK8sNamespacesSet) {
 		return ctrl.Result{}, nil
 	}
 
-	err := r.Client.Get(ctx, req.NamespacedName, &serviceEndpoints)
+	err := r.Client.Get(ctx, req.NamespacedName, &serviceEndpointSlice)
 
-	// endpointPods holds a set of all pods this endpoints object is currently pointing to.
+	// endpointPods holds a set of all pods this endpoint slice object is currently pointing to.
 	// We use this later when we reconcile ACL tokens to decide whether an ACL token in Consul
 	// is for a pod that no longer exists.
 	endpointPods := mapset.NewSet()
 
-	// If the endpoints object has been deleted (and we get an IsNotFound
+	// If the endpoint slice object has been deleted (and we get an IsNotFound
 	// error), we need to deregister all instances in Consul for that service.
 	if k8serrors.IsNotFound(err) {
 		// Deregister all instances in Consul for this service. The function deregisterServiceOnAllAgents handles
```

Comment:
If my service has 2000 pods then I'll have >= 1 endpoint slice. If I scale that down to 100 pods then I'll have only one endpoint slice. At that point we'll get an event hat the endpoint slice has been deleted and we'll run through this function.

Now we'll attempt to deregister _all_ instances of this service but there's still 100 running.

Reply:
Maybe we should have a meta tags on all the service registrations which maps to the name of the endpoint-slice object? We could use that to determine which service instances to deregister.

Reply:
Is each pod guaranteed to always be on the same ep slice?